### PR TITLE
GetResourceName default parameter and Global variable

### LINKED ISSUE #3527: Map Editor bugs out
### Describe the bug

Hello there!

Recently noticed a strange behaviour from Map Editor. My friend and other friends also got it. It started when MTA started launching weirdly. Yes I do have ASI Mods but they never affected the game, and usually prompt about "Installed ASI" things popped up fast. Recently however, when I launched MTA:SA it showed the logo, but then without any prompt about installing new updates or something icon dissapeared 3 times and it launched. It's still going to this time, but at that time still editor was fine. Yesterday editor was also fine for me, however today I wanted to do a new RACE map and it half-worked. Checkpoints didn't had any arrows when I put them and when you usually select the Checkpoint by just clicking on it, and duplicating it - it already sets nextid for the next correct checkpoint. It didn't happen. I needed to manually re-add nextid to every checkpoint but that's just tip of the iceberg. When I quit, the worse thing started to happen that my friends and others report to me. As soon as I go to the editor i have problems:

[17:23:18] WARNING: editor_main\server\mapsettingssync.lua:158: The resource isn't active
[17:23:18] ERROR: editor_main\server\mapsettingssync.lua:159: bad argument #1 to 'lower' (string expected, got boolean)
[17:23:24] WARNING: editor_main\server\mapsettingssync.lua:158: The resource isn't active

What the hell is happening? I clearly have all gamemodes in, but in the next screenshot im about to show you there are no gamemodes, no nothing. I can't ever save and it says at the beginning that editor_dump is missing (it randomly vanished, wtf?)

[17:34:41] ERROR: editor_main\server\mapsettingssync.lua:159: bad argument #1 to 'lower' (string expected, got boolean)
mta-screen_2024-06-30_17-34-44
mta-screen_2024-06-30_17-34-51

Can you please explain, what might be happening? I don't want to reinstall my MTA yet again. And strangely enough, when my friend reinstalled MTA problem came back.

### Steps to reproduce

- Just simply enter the editor
- In my case I did a RACE map in the Unity Station with Glendale - 22 checkpoints
- I saved the map, everything is fine up to this point
- I tested the map, it worked fine
- I quit the editor
- I quit the game
- I came back to the game
- Opened editor again
- The error pops up about editor_dump
- And first thing I see in the console is 

- I try to make a new map - nothing happens, nothing loads
- I try to open Map Settings / Definitions / in Map Settings / Gamemodes (nothing is there)
- I try to load the map (nothing works)

### Version

Multi Theft Auto: San Andreas v1.6

### Additional context

_No response_

### Relevant log output

_No response_

### Security Policy

- [X] I have read and understood the Security Policy and this issue is not security related.

### PR DESCRIPTION
**Pull request by ghost**
### Additions
- Adds a default parameter to the `getResourceName` function, which is based on the current resource
- Adds a global variable named `resourceName`, which is based on the current resource

### Antecedent
This is a continuation of the previous pull request (#3363) which fixed the previously mentioned map editor issues.
Thus this pull request closes #3527

### Example
```lua
addEventHandler("onResourceStart", resourceRoot,
    function()
        local myResourceName = resourceName or getResourceName()
        local myResourceMessage = string.format("Resource successfully started: %s", myResourceName)

        outputServerLog(myResourceMessage)
    end
)
```

### CODE REVIEW

**Comment by Dutchman101** [comment]
Both you and Tracer submitted a fixed PR at the exact same time (lol)

Code reviews and (self-)feedback should determine which of the 2 PR's is ideal, and to save us time, also please add details on what the problem was and how you fixed it; there's no specific "fix" commit.

Tracer's PR: https://github.com/multitheftauto/mtasa-blue/pull/3530
Nico's PR: https://github.com/multitheftauto/mtasa-blue/pull/3531

**Comment by ghost** [comment]
This pull request contains the option to omit the parameter of the `getResourceName` function, which was the intended functionality of the reverted pull request, so this pull request restores it with the correction.
The problem was with the `!resource->IsActive()` check, after removing this the mentioned issue no longer occurs.

cc @Dutchman101

**Comment by ghost** [comment]
The other pull request was closed by Tracer's own decision, so this is the only one left.

File: Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp [hunk 1]
```diff
@@ -220,34 +222,22 @@ int CLuaResourceDefs::GetResourceConfig(lua_State* luaVM)
...
 }
 
-int CLuaResourceDefs::GetResourceName(lua_State* luaVM)
+std::string CLuaResourceDefs::GetResourceName(lua_State* luaVM, std::optional<CResource*> resourceElement)
 {
-    // Verify arguments
-    CResource*       pResource = NULL;
-    CScriptArgReader argStream(luaVM);
-    argStream.ReadUserData(pResource);
+    if (resourceElement && resourceElement.has_value())
```
Comment (TracerDS) [inline Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: just `resourceElement.has_value()` would be sufficient

**Review by TracerDS — APPROVED** [review]
👍

File: Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp (hunk 1 above)
Comment (ghost) [reply Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: Thx

**Review by TracerDS — APPROVED** [review]

**Review by FileEX — APPROVED** [review]

File: Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp [hunk 2]
```diff
@@ -898,23 +898,22 @@ int CLuaResourceDefs::getResourceLoadTime(lua_State* luaVM)
...
-int CLuaResourceDefs::getResourceName(lua_State* luaVM)
+std::string CLuaResourceDefs::GetResourceName(lua_State* luaVM, std::optional<CResource*> resourceElement)
 {
-    CResource* pResource;
+    if (resourceElement.has_value())
+        return (*resourceElement)->GetName();
 
-    CScriptArgReader argStream(luaVM);
-    argStream.ReadUserData(pResource);
+    CLuaMain* localVM = m_pLuaManager->GetVirtualMachine(luaVM);
```
Comment (FileEX) [inline Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: You can do?
```cpp
CResource* resource = m_pLuaManager->GetVirtualMachineResource(luaVM);
```

**Review by FileEX — CHANGES_REQUESTED** [review]

File: Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp (hunk 2 above)
Comment (Lpsd) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: Use `lua_getownercluamain` instead
```cpp
CLuaMain& luaMain = lua_getownercluamain(luaVM);
CResource* localResource = luaMain.GetResource();
```

You can even shorten to `CResource* localResource = lua_getownercluamain(luaVM).GetResource();`

Comment (ghost) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: Thx

File: Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp [hunk 3]
```diff
@@ -220,34 +222,18 @@ int CLuaResourceDefs::GetResourceConfig(lua_State* luaVM)
...
-                return 1;
-            }
-        }
-        else
-            m_pScriptDebugging->LogBadPointer(luaVM, "resource", 1);
-    }
-    else
-        m_pScriptDebugging->LogCustom(luaVM, argStream.GetFullErrorMessage());
+    CLuaMain&  luaMain = lua_getownercluamain(luaVM);
+    CResource* localResource = luaMain.GetResource();
```
Comment (TracerDS) [inline Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]:
```cpp
CResource* localResource = &lua_getownerresource(luaVM);
```

File: Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp [hunk 4]
```diff
@@ -898,23 +898,18 @@ int CLuaResourceDefs::getResourceLoadTime(lua_State* luaVM)
...
+std::string CLuaResourceDefs::GetResourceName(lua_State* luaVM, std::optional<CResource*> resourceElement)
 {
-    CResource* pResource;
+    if (resourceElement.has_value())
+        return (*resourceElement)->GetName();
 
-    CScriptArgReader argStream(luaVM);
-    argStream.ReadUserData(pResource);
+    CLuaMain&  luaMain = lua_getownercluamain(luaVM);
+    CResource* localResource = luaMain.GetResource();
```
Comment (TracerDS) [inline Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]:
```cpp
CResource* localResource = &lua_getownerresource(luaVM);
```

**Review by TracerDS — CHANGES_REQUESTED** [review]

File: Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp (hunk 3 above)
Comment (ghost) [reply Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: Thx

File: Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp (hunk 4 above)
Comment (ghost) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: Thx

**Review by TracerDS — APPROVED** [review]
LGTM

**Pull request merged**